### PR TITLE
[SPARK-31719][SQL] Refactor JoinSelection

### PR DESCRIPTION
**Pull request by dbaliafroozeh**
### What changes were proposed in this pull request?
This PR extracts the logic for selecting the planned join type out of the `JoinSelection` rule and moves it to `JoinSelectionHelper` in Catalyst.

### Why are the changes needed?
This change both cleans up the code in `JoinSelection` and allows the logic to be in one place and be used from other rules that need to make decision based on the join type before the planning time.

### Does this PR introduce _any_ user-facing change?
`BuildSide`, `BuildLeft`, and `BuildRight` are moved from `org.apache.spark.sql.execution` to Catalyst in `org.apache.spark.sql.catalyst.optimizer`.

### How was this patch tested?
This is a refactoring, passes existing tests.

### CODE REVIEW

**Comment by hvanhovell** [comment]
ok to test

**Comment by hvanhovell** [comment]
add to whitelist

**Comment by maropu** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala [hunk 1]
```diff
@@ -208,3 +208,161 @@ object ExtractPythonUDFFromJoinCondition extends Rule[LogicalPlan] with Predicat
...
   }
 }
+
+sealed abstract class BuildSide
+
+case object BuildRight extends BuildSide
+
+case object BuildLeft extends BuildSide
+
+trait JoinSelectionHelper {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:218]: We need to pull out this functionality as a helper trait? For example, `PredicateHelper` is used in many places, but this helper trait is only used for `JoinSelection `?

Comment (dbaliafroozeh) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:218]: @maropu That's true that currently there are no other uses of this trait, but there are potentially other rules that can use the join selection logic at optimization time, so this refactoring will open up the opportunity for more optimization rules.
Also, this cleans up the JoinSelection rule quite a bit, and IMO the logic for determining the type of the join is more clear now.

**Comment by dbaliafroozeh** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala (hunk 1 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:218]: okay, that sounds fine to me. btw, if the helper is more general as you said, how about `JoinSelectionHelper` -> `JoinPlanHelper `?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:218]: Also, could you add some tests for this new helper?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala [hunk 2]
```diff
@@ -208,3 +208,161 @@ object ExtractPythonUDFFromJoinCondition extends Rule[LogicalPlan] with Predicat
...
+
+  def getBroadcastBuildSide(
+      left: LogicalPlan,
+      right: LogicalPlan,
+      joinType: JoinType,
+      hint: JoinHint,
+      onlyLookingAtHint: Boolean,
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:225]: nit: `selectOnlyFromHint`? I couldn't tell what this meant just by the name `onlyLookingAtHint `.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:225]: If we want to be super clear, maybe "pickJoinSideByHintOnly"? or simply "hintOnly".

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:225]: Ah, `hintOnly` looks nice to me.

**Review by cloud-fan — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Review by maropu — APPROVED** [review]
LGTM if the tests passed.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by cloud-fan** [comment]
oops, the latest commit hasn't passed tests yet. I'll monitor the jenkins status.

**Pull request closed without merging**